docs/source/add_new_model.md

```diff
@@ -0,0 +1,573 @@
+  cd brand_new_bert
+  pip install -e .
+  ```
+
+Now you have set up a development environment to port brand_new_bert to 🤗Transformers.
+  
+### Run a pretrained checkpoint using the original repository
+
+At first, you will work on the original repository. Often, the original implementation is very
+ "researchy" meaning that documentation might be lacking and the code can be difficult to understand.
+But this should be exactly your motivation to reimplement brand_new_bert. At Hugging Face, one of our main goals is to
+*make people stand on the shoulders of giants* which translates here very well into taking a working 
+model and rewriting it to make it as **accessible, user-friendly, and beautiful** as possible.
+This is the #1 motivation to re-implement models into 🤗Transformers - trying to maximize access 
+to a complex new NLP technology for **everybody**.
+	
+You should start thereby by diving into the original repository.
+
+#### Get familiar with the original repository.
+
+Successfully running the official pre-trained model in the original repository is often 
+**the most difficult** step. From our experience, it is very important to spend some time to get familiar with the original codebase. You should find out
+
+- Where to find the pre-trained weights
+- How to load the pre-trained weights into its corresponding model
+- Trace one forward pass so that you know which classes and functions are required for a simple forward pass
+  . Usually, you only have to reimplement those functions.
+- Be able to locate the important components of the model: Where is the model class? Are there submodel 
+  classes, *e.g.* EncoderModel, DecoderModel? Where is the self-attention layer? 
+  Are there multiple different attention layers, *e.g.* *self-attention*, *cross-attention*...?
+- How can you debug the model in the original environment of the repo? Do you have to set `print` statements
+  or can you work with an interactive debugger like `ipdb`?
+  
+It is very important that before you start opening a PR in 🤗Transformers that you can **efficiently** 
+debug code in the original repository! This means that you should be able to run a forward pass and print out the actual values of the output of a layer. *I.e* you can load a pre-trained model, pass an input vector of 
+token ids, *i.e.* `input_ids = [0, 1, 4, 5, ...]` to the model's forward function and you can print out the 
+intermediate outputs of - let's say - the first self-attention layer that could look something like this: 
+
+```bash
+[[
+ [-0.1465, -0.6501,  0.1993,  ...,  0.1451,  0.3430,  0.6024],
+ [-0.4417, -0.5920,  0.3450,  ..., -0.3062,  0.6182,  0.7132],
+ [-0.5009, -0.7122,  0.4548,  ..., -0.3662,  0.6091,  0.7648],
+ ...,
+ [-0.5613, -0.6332,  0.4324,  ..., -0.3792,  0.7372,  0.9288],
+ [-0.5416, -0.6345,  0.4180,  ..., -0.3564,  0.6992,  0.9191],
+ [-0.5334, -0.6403,  0.4271,  ..., -0.3339,  0.6533,  0.8694]]],
+```
```

Comment:
While printing can get things done, I think there are better ways to go about it, which may be a bit more complex to setup, but end up being **so** useful as development goes on.

When porting a model, I like to:
- Isolate the different forward passes that are useful, in a step-by-step manner. This means that I like to proceed as follows:
    - Have a way to retrieve the embeddings
    - Have a way to retrieve each layer's output
    - Have a way to retrieve the Transformer output
    - Have a way to retrieve the full model output, for each head
- All of these output should be gathered into a single script, in **variables**. This can be a pain to setup, but makes it so much simpler in the end because you can **compare** values.
- These values should be compared to the implementation in HuggingFace Transformers. A relative/absolute difference can be defined

Having this step-by-step approach with the ability to compare is imo incredibly useful because:
- You can check the results algorithmically, rather than checking the terminal output and verifying the results
- You can set this up very early on, starting with the embeddings
- You can ensure you have no regression when working on the model, as when finishing up the last head, you'll still be checking the embeddings. You'll be checking all heads at once, too.

Here's an example of how I did it with ELECTRA: [compare_electra.py](https://gist.github.com/LysandreJik/db4c948f6b4483960de5cbac598ad4ed). It's slightly more complicated to setup but I think it's worthwhile.

I have no idea how you would explain this in the document, so feel free to ignore if it's not appropriate :)

Reply:
Yes, I think that's the more in-depth approach! I did this for Reformer as well. Maybe write down both approaches?

Reply:
I personally prefer to have everything in a notebook where I develop new components and import old components to check the outputs of every single object match. I think we should also show that approach, with some "choose your own adventure" question: are you more comfortable with notebooks or scripts? In case a..., in case b...